elastic/beats
# [Metricbeat] Add tags as a filter for cloudwatch metricset

This PR is to add a `tags` filter in cloudwatch configuration to allow users only collect metrics from instances/services that have the tags specified in the config. 
Adding this feature turned out to require a lot bigger _refactoring_ of the old code than I thought.

For example:
```
- module: aws
  period: 300s
  metricsets:
    - cloudwatch
  metrics:
    - namespace: AWS/EC2
      tags.resource_type_filter: ec2:instance
      statistic: ["Average"]
      tags:
        - key: "Name"
          value: "ECS Instance test"
```
With this configuration, cloudwatch metricset should only collect metrics from EC2 instances which have tag `Name: ECS Instance test`. 

If there are more than one set of tags specified, then all the tags should match. For example:
```
- module: aws
  period: 300s
  metricsets:
    - cloudwatch
  metrics:
    - namespace: AWS/EC2
      tags.resource_type_filter: ec2:instance
      statistic: ["Average"]
      tags:
        - key: "Name"
          value: "ECS Instance test"
        - key: "Organization"
          value: "Engineering"
```
With this configuration, cloudwatch metricset should only collect metrics from EC2 instances which have both tag `Name: ECS Instance test` and tag `Organization: Engineering`. 

Maybe in the future, other features can be added for tag filtering, such as supporting regular expressions in tag filters.

This implements a part of https://github.com/elastic/beats/issues/13145

### How to test this
Note: The tags used in configuration examples below might not work for your aws account. Please make sure to either use tags exist from instances from your aws account or create these tags for testing purpose only. 

1. Test with collecting tags and no tags
```
- module: aws
  period: 300s
  metricsets:
    - cloudwatch
  credential_profile_name: test-mb
  metrics:
    - namespace: AWS/EC2
      name: ["CPUUtilization", "DiskWriteOps"]
      tags.resource_type_filter: ec2:instance
      statistic: ["Average"]
      tags:
        - key: "Name"
          value: "ECS Instance - DockerRegistryECSStack"
    - namespace: AWS/ELB
```
With this configuration, you should see EC2 `aws.metrics.CPUUtilization.avg` and `aws.metrics.DiskWriteOps.avg` metrics but only with tag `Name="ECS Instance - DockerRegistryECSStack"`. 
Also all metrics from ELB with no tags.

2. Test with more than one set of tags
```
- module: aws
  period: 300s
  metricsets:
    - cloudwatch
  credential_profile_name: test-mb
  metrics:
    - namespace: AWS/EC2
      name: ["CPUUtilization", "DiskWriteOps"]
      tags.resource_type_filter: ec2:instance
      statistic: ["Average"]
      tags:
        - key: "cluster"
          value: "bastion-hosts"
        - key: "managed-by"
          value: "ansible"
```
With this configuration, you should see EC2 `aws.metrics.CPUUtilization.avg` and `aws.metrics.DiskWriteOps.avg` metrics but only with tag `cluster="bastion-hosts"` and tag `managed-by="ansible"`.

3. Test with dimensions specified
```
- module: aws
  period: 300s
  credential_profile_name: test-mb
  metricsets:
    - cloudwatch
  metrics:
    - namespace: AWS/EC2
      name: ["CPUUtilization"]
      tags.resource_type_filter: ec2:instance
      dimensions:
        - name: InstanceId
          value: i-0987d3e7a753a8dab
      statistic: ["Average", "Maximum"]
```
With this configuration, you should see EC2 `aws.metrics.CPUUtilization.avg` and `aws.metrics.CPUUtilization.max` metrics for only `InstanceId = i-0987d3e7a753a8dab` with tags.
